### ISSUE TITLE
20.09 Feature freeze items

### ISSUE THREAD

**Issue by jonringer** [NixOS/nixpkgs#95475]
Pinging all language, framework, and ecosystem owners to consolidate feature freeze items for the 20.09 release.

Please mention any items you see blocking the 20.09 release in your given domains

Items brought up will be discussed in a Feature freeze meeting at a later date.

Nix/nix-cli ecosystem: @edolstra  @grahamc  @nbp @Profpatsch 
Mobile: @samueldr 
Nixos Modules / internals : @Infinisil @matthewbauer @Ericson2314 @orivej 
Nixos tests: @tfc

Emacs: @adisbladis
Erlang: @gleber 
Go: @kalbasit @Mic92 @zowoq
Haskell: @peti @cdepillabout 
Python: @FRidh 
Perl: @volth 
php: @NixOS/php
Ruby: @alyssais 
rust: @bhipple @Mic92 @andir @LnL7 

Darwin: @NixOS/darwin-maintainers 

bazel: @mboes
blockchains @mmahut
podman: @NixOS/podman
Qt / KDE: @ttuegel
Postgres: @thoughtpolice 

Co-release-manager: @worldofpeace 

in case I forgot anyone: @NixOS/nixpkgs-committers 

If I did forget someone, please cc them. I would like to make sure that everyone was able to voice their concerns and issues.

**Comment by cdepillabout** [comment]
cc @maralorn about Haskell stuff as well.

**Comment by vcunat** [comment]
Roughly it should be master around the end of August as usual, right?

**Comment by jonringer** [comment]
Our current plan is to have the branch-off + ZHF date on 4 Sept. I'll make a more official roadmap announcement tomorrow.

**Comment by arianvp** [comment]
#94354 is blocking

**Comment by volth** [comment]
* perl: 5.30.x -> 5.32.x (https://github.com/NixOS/nixpkgs/pull/95012 + https://github.com/NixOS/nixpkgs/pull/93187)
* coming mass update of `perlPackages` after `perl` upgrade

**Comment by aanderse** [comment]
Looks like `zabbix.server-mysql` and `zabbix.server-pgsql` broke recently: https://hydra.nixos.org/build/122059780. I briefly poked around but I'm not sure what the problem or solution is. If anyone has any insight much appreciated.

**Comment by jonringer** [comment]
@aanderse It tries to compile a small program against `libxml2`, but fails to do so; although the `PKG_CONFIG_PATH` appears to contain libxml2.

It can be fixed short term by removing `libxml2`, but I'm not sure if that's acceptable for the server

**Comment by etu** [comment]
Just to mention it as a representative/maintainer of the php ecosystem. We've recently merged our deprecation of on old version that should be merged before 20.09. So for the php ecosystem we're good to go.

Consider this message a confirmation that someone read the ping to that team in the posting.

**Comment by bhipple** [comment]
Not aware of any major outstanding blocking issues in the Rust ecosystem.  The `staging` branch moves `rustc 1.45.0 -> 1.45.2`, which is the latest; I'm assuming we'll get at least 1 more staging -> master merge before branch-off.

**Comment by jonringer** [comment]
> I'm assuming we'll get at least 1 more staging -> master merge before branch-off.

That should be the case

**Comment by aanderse** [comment]
> @aanderse It tries to compile a small program against `libxml2`, but fails to do so; although the `PKG_CONFIG_PATH` appears to contain libxml2.
> 
> It can be fixed short term by removing `libxml2`, but I'm not sure if that's acceptable for the server

@jonringer unfortunately that would remove functionality (automatic vmware monitoring) that some users **may** find important. I personally don't use that functionality, so **I** would accept that over a broken build... I have opened a separate [issue](https://github.com/NixOS/nixpkgs/issues/95532) to track this.

**Comment by aanderse** [comment]
Another issue which needs resolution one way or the other: https://github.com/NixOS/nixpkgs/pull/94022.

ping @grahamc @flokli @arianvp @andir @blitz @primeos

**Comment by worldofpeace** [comment]
According to the GNOME 3.38.0 roadmap https://wiki.gnome.org/ThreePointThirtyseven it's due Sep 16. So I believe this means we won't have the latest GNOME in the release. This is fine since we basically never do because our release time conflicts with theirs always.

**Comment by worldofpeace** [comment]
It was projected that xfce 4.16 would be available before 20.09 release, but it seems they need more time in their roadmap https://github.com/NixOS/nixpkgs/issues/81569. So xfce 4.16 will not be in 20.09.

**Comment by bjornfor** [comment]
> So I believe this means we won't have the latest GNOME in the release. This is fine since we basically never do because our release time conflicts with theirs always.

Why don't we release NixOS one month later? This will sync up with Ubuntu too.

**Comment by vcunat** [comment]
Could be, but there are many projects that we package, and so far I haven't noticed that a notable fraction of them synchronizes around some dates (like the Ubuntu schedule).

**Comment by Atemu** [comment]
@worldofpeace do you know how big the GNOME breakage would be?  
If it's not too bad I don't see why we couldn't just backport it after release.

**Comment by vcunat** [comment]
Generally I wouldn't do such things, and for desktops in particular I don't anticipate sufficient motivation for doing that.  Regardless of release dates, there's a notable advantage we have in comparison to the Ubuntu model: we do provide quite a usable rolling version.  People that want the latest _whatever_ can use `nixos-unstable` channel (which still gives some automatic-testing assurances).

**Comment by jonringer** [comment]
> Why don't we release NixOS one month later? This will sync up with Ubuntu too.

I'm not sure if that's a good reason to push a date.  Unless KDE, gnome, and the other big software suites decides to align themselves for that type of schedule, then may it would make sense for us to change so we could have the "latest" going into a release.

However, it also takes time for downstream packages to fully support those versions, and it still takes time to ensure they are working as intended. It's going to be hit or miss whether a given month really affords any substantial opportunities given there's so many packages, and not everyone cares certain sets. For example, I don't use a DE, so I largely am unaffected by KDE, plasma, gnome, or other changes. 

And as @vcunat has mentioned, we have a _very_ usable "unstable" channel if people really care about having the latest and greatest.

And one thing NixOS can do which other distros usually cannot, is allow for a user to pick packages from different channels. (although, this is much more involved for a regular user)

**Comment by Mic92** [comment]
> 
> And one thing NixOS can do which other distros usually cannot, is allow for a user to pick packages from different channels. (although, this is much more involved for a regular user)

This become less involved with flakes.

**Comment by worldofpeace** [comment]
> @worldofpeace do you know how big the GNOME breakage would be?

The amount of breakage when updating gnome can vary, but in general updating desktop environments etc. in nixos is very difficult because we don't really have an automatic testing to verify.  Here's an excellent example of this issue https://github.com/NixOS/nixpkgs/issues/84634#issuecomment-674729990. There are bugs we haven't noticed for a while because it's a thing I don't use.
So I'm actually thankful we'll have 3.36 on stable, which is something that has been checked up on for a while now. 

> If it's not too bad I don't see why we couldn't just backport it after release.

That is a major upgrade, and generally, in a stable distribution, it is something that couldn't be backported.
After 20.09 is released it's totally "feature frozen". So that would be a bunch of UI/UX changes breaking the freeze.

> Why don't we release NixOS one month later? This will sync up with Ubuntu too.

As for ever pushing our release a month forward to accommodate gnome's release (which happens in the *.03 in *.09 exactly btw), this still wouldn't be enough time because we lack the resources. I would agree with most things @jonringer mentioned https://github.com/NixOS/nixpkgs/issues/95475#issuecomment-674690777, but I do understand for a general user pov accommodating to have the latest just makes sense. I think changing the actual dates in which we "consistently try to aim for" should be done with a lot of reasoning.  Though let's be real: NixOS 20.03 was a NixOS 20.04. But pushing the date to the month we actually are releasing on isn't going to make us more punctual either :rofl: .
Anyways, I think this a discussion better had at https://github.com/NixOS/nixpkgs/issues/95535. 
My principle for changing things up is we should do smaller but still impacting and meaningful changes. See https://github.com/NixOS/nixpkgs/issues/95535 (Changes from last year). We have to remember that someone new is going to have to duplicate this process in six months and still produce a consistent experience.

**Comment by adisbladis** [comment]
I see https://github.com/NixOS/nixpkgs/pull/87268 as blocking. I would love some input on what a good solution is (in the PR).

**Comment by NeQuissimus** [comment]
I'd like to see the systemd updates from staging make it into master, so we can bring back the 5.8.x kernel.

**Comment by vcunat** [comment]
That one is very certain.

**Comment by mweinelt** [comment]
I wonder what the @NixOS/acme team thinks about supporting the current ACME module for another release, when it would probably be best to be investing in https://github.com/NixOS/nixpkgs/pull/91121.

**Comment by m1cr0man** [comment]
> I wonder what the @NixOS/acme team thinks about supporting the current ACME module for another release, when it would probably be best to be investing in #91121 asap.

I for one would very much like to see it merged. Maintaining the "old" module continues to get more painful as more patches and bug fixes are added to address legacy behaviour. I've addressed all the issues I could find in Github and elsewhere, and having had it in production on my personal system for quite a while I haven't had a single renewal issue.

**Comment by jonringer** [comment]
To avoid spamming everyone with yet another issue, please add release items that are not in the current 20.09 release notes in the issue: #95765

**Comment by Atemu** [comment]
Shouldn't we rather fix the image by putting the RPI kernel Fork in it instead of mainline?

**Comment by samueldr** [comment]
> Shouldn't we rather fix the image by putting the RPI kernel Fork in it instead of mainline?

Putting the RPI kernel fork in what? This is what the Raspberry Pi 4-specific image does already. The task is to remove the undesirable device-specific image from the stable channel.

The sd image, by design, is supposed to be generic. The current Raspberry Pi 4-specific sd image derivation only exists because of the popularity of the platform, and demand. I still hold the opinion that it might have been a mistake to have it at all. If the platform can't boot using the standard boot methods (u-boot) and the mainline kernel, for the time being it's not worth the headaches of maintaining it in an official manner.

**Comment by jonringer** [comment]
I'm unpinning this as many of the items have been merged in before branch off, if they have not please add them to the [20.09 blocker project](https://github.com/NixOS/nixpkgs/projects/33)

**Comment by jonringer** [comment]
I'm going to re-pin this, as many items have been added, but few have seen progress in the past 3 weeks.

Items that feel like they warrant delaying the entirety of the release please make a comment here with:
- Name or summary of issue
- Related PR or issue
- (Optional, if impact isn't denoted by name or summary) Impact on nixos users
- (Recommended) Timeline of when issue can be resolved or mitigated.

NixOS is a community of volunteers, let's reasonable about time exceptions on getting these issues resolved. As much as I would love to have all scenarios be a tier 1 experience, the reality is that it's just not feasible.

Scenarios involving casual users with "vanilla" (E.g. plasma, gnome, other DEs) desktop needs and production users with security needs (E.g. postgresql) are the two big user bases that I think warrant delaying the release. But the purpose of re-activating this thread is to open a forum for discussing such issues.

**Comment by volth** [comment]
Do not forget https://github.com/NixOS/nixpkgs/pull/97522, currently perl is broken in 20.09 branch after 5.30 -> 5.32 update, so the packages must be updated too (it is done in master but after 20.09 spin-off)

**Comment by OPNA2608** [comment]
Unsure where to mention this, but would `clangMultiStdenv` / `cc-multilib-test` on Hydra being broken since July be a blocker or otherwise noteworthy for the release notes? See https://github.com/NixOS/nixpkgs/issues/94023, https://hydra.nixos.org/job/nixpkgs/staging-next/tests.cc-multilib-clang.x86_64-linux/

**Comment by mweinelt** [comment]
In https://github.com/NixOS/nixpkgs/pull/100195 redis default bind is readjusted to localhost, which corrects what to me seems like unexpected (no other distro does this) and insecure behaviour (no auth by default) and I'd be glad if we could get that into 20.09.

**Comment by worldofpeace** [comment]
I wanted to land this feature, even though I know its dead last minute https://github.com/NixOS/nixpkgs/pull/100199

**Comment by rissson** [comment]
In my mind #101099 is a blocker. Sorry if that was the wrong issue to post this.

**Comment by jonringer** [comment]
This issue has served it's purpose. With #101736 merged, I think this is good to close
